mendix/cf-mendix-buildpack
# Encrypt connections between App & RDS PostgreSQL instance

To satisfy requirements of PCI, a connection to a DB instance must be encrypted with SSL/TLS.

With this change JDBC connection to a PostgreSQL database instance running on AWS RDS will be encrypted using `verify-full` option by default. To identify if we are running on AWS RDS database host parameter is matched against "*.rds.amazonaws.aws" pattern.

> In verify-full mode, the host name is matched against the certificate's Subject Alternative Name attribute(s), or against the Common Name attribute if no Subject Alternative Name of type dNSName is present.

DatabaseJdbcUrl parameter for PostgreSQL databases only supported by Mendix ≥ 7, so this change will enable SSL/TLS encryption only for the applications that are running with Mendix ≥ 7.
https://github.com/mendix/docs/blob/d23111e40e34b3aa8d13943745f150e09c0efee0/content/refguide6/custom-settings.md#41-common-settings

Additionally configure PostgreSQL JDBC driver to use default java SSLFactory as a certificate source.
https://jdbc.postgresql.org/documentation/head/ssl-factory.html

 ---

Manual tests:

<details><summary>1. Without the RDS CA 2019 certificate in Java Keystore</summary>
<p>

1. Push an app:

    ```bash
    ❯ cf push -f test.yml -p mendix7/AdoptOpenJDKTest_7.23.1.mda
    ...
    INFO: Preflight check on Mendix runtime version [7.23.1.46759] and stack [cflinuxfs3]...
    ...
    ```

1. Verify that the application cannot start without the correct certificate:

    ```bash
    ❯ cf logs test-ssl
    ...
    ERROR - ConnectionBus: Opening JDBC connection to Some(dbcbz2r8d9vtewzzk.cgmvyifrogdx.eu-central-1.rds.amazonaws.com:5432) failed with SQLState: 08006 Error code: 0 Message: SSL error: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target Retrying...(1/4)
    ...
    Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
    ```

</p>
</details>

<details><summary>2. Mendix 7</summary>
<p>

1. Push an app:

    ```bash
    ❯ cf push -f test.yml -p mendix7/AdoptOpenJDKTest_7.23.1.mda
    ...
    INFO: Preflight check on Mendix runtime version [7.23.1.46759] and stack [cflinuxfs3]...
    ...
    ```

1. Verify that SSL parameters are present in JDBC connection string:

    ```bash
    ❯ cf logs test-ssl
    ...
    2020-08-26T12:00:03.48+0200 [APP/PROC/WEB/0] OUT INFO - ConnectionBus: Database: PostgreSQL 9.6.17, name: 'jdbc:postgresql://dbs618asx9jkyt44m.cgmvyifrogdx.eu-central-1.rds.amazonaws.com:5432/dbs618asx9jkyt44m?tcpKeepAlive=true&sslmode=verify-full&sslfactory=org.postgresql.ssl.DefaultJavaSSLFactory'
    ```

1. Verify client address for the SSL connections:

    ```bash
    dbs618asx9jkyt44m=> select * from pg_stat_ssl;
      pid  | ssl | version |           cipher            | bits | compression | clientdn
    -------+-----+---------+-----------------------------+------+-------------+----------
      5974 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
     13483 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
     13494 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
     13524 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
     15047 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
    (5 rows)

    dbs618asx9jkyt44m=> SELECT client_addr, query FROM pg_stat_activity where pid = '13494';
     client_addr  |                               query
    --------------+-------------------------------------------------------------------
     10.10.136.20 | SELECT COUNT(id) from system$filedocument WHERE hascontents=true;
    (1 row)

    ❯ cf curl /v2/apps/$(cf app test-ssl --guid)/stats | jq '."0".stats.host'
    "10.10.136.20
    ```

</p>
</details>

<details><summary>3. Mendix 8</summary>
<p>

1. Push an app:

    ```bash
    ❯ cf push -f test.yml -p mendix8/Mendix8.1.1.58432_StarterApp.mda
    ...
    INFO: Preflight check on Mendix runtime version [8.1.1.58432] and stack [cflinuxfs3]...
    ...
    ```

1. Verify that SSL parameters are present in JDBC connection string:

    ```bash
    ❯ cf logs test-ssl
    ...
    2020-08-26T12:21:56.49+0200 [APP/PROC/WEB/0] OUT INFO - ConnectionBus: Database: PostgreSQL 9.6.17, name: 'jdbc:postgresql://db1e92nqwzi6zzqm5.cgmvyifrogdx.eu-central-1.rds.amazonaws.com:5432/db1e92nqwzi6zzqm5?tcpKeepAlive=true&sslmode=verify-full&sslfactory=org.postgresql.ssl.DefaultJavaSSLFactory'
    ```

1. Verify client address for the SSL connections:

    ```bash
    db1e92nqwzi6zzqm5=> select * from pg_stat_ssl;
      pid  | ssl | version |           cipher            | bits | compression | clientdn
    -------+-----+---------+-----------------------------+------+-------------+----------
      6148 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
      9628 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
      9632 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
      9636 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
     10352 | t   | TLSv1.2 | ECDHE-RSA-AES256-GCM-SHA384 |  256 | f           |
    (5 rows)

    db1e92nqwzi6zzqm5=> SELECT client_addr, query FROM pg_stat_activity where pid = '9636';
     client_addr  |                               query
    --------------+-------------------------------------------------------------------
     10.10.136.20 | SELECT COUNT(id) from system$filedocument WHERE hascontents=true;
    (1 row)

    ❯ cf curl /v2/apps/$(cf app test-ssl --guid)/stats | jq '."0".stats.host'
    "10.10.136.20"
    ```

</p>
</details>
